design/design-doc-composition-functions-extra-resources.md

```diff
@@ -0,0 +1,1196 @@
+  request?
+  - **Q5.1**: When should the policy be evaluated?
+    - **Option 1 (Q5.1)**: Before gathering the requested extra resources.
+  - **Q5.2**: What should be the policy's input?
+    - **Option 1 (Q5.2)**: The request only.
+- **Q6**: What about Resolve and Resolution Policies?
+  - Decided not to allow configuring it, defaulting to `resolve` policy `Always` and
+    `resolution` policy left to Functions to implement.
+- **Q7**: How can we support "Realtime Compositions"?
+  - **Q7.1**: How should Functions specify what resources should be watched?
+    - **Option 2 (Q7.1)**: Let Functions specify which resources should be watched.
+  - **Q7.2**:  How should Functions specify what resources should be watched?
+    - **Option 2 (Q7.2)**: Functions request to watch specific resources separately
+      and only by name.
+  - **Q7.3**: Where should Crossplane store the references to the extra resources
+    requested by Functions?
+    - **Option 3 (Q7.3)**: In the XR's `status`, in a new dedicated field.
+  - **Q7.4**: What resources should Functions be able to watch?
+    - **Option 2 (Q7.4)**: Only resources requested in previous iterations.
+
+### Q0: Should Functions be able to request additional resources?
+
+Clearly this is a yes or no question, but we need to do a step back. What we
+actually care about is to have a way to allow **Composition authors** to specify
+additional resources Functions should have access to. So, the options we have
+are actually:
+- **Option 1**: Yes, Composition authors should rely on Functions to request
+  additional resources.
+- **Option 2**: No, Composition authors should not need Functions at all to
+  request additional resources, it should be part of the Composition API.
+
+**Option 2** would feel more "native", everything would be handled by
+Crossplane, and we could design an API up to our standards. Ignoring native
+patch-and-transform Compositions for now, which would present their own set of
+issues around handling multiple results, due to the limited expressiveness of
+the downstream language (P&T). Even if we decided to make it only available for
+Functions based compositions, the API would need to be pretty expressive to
+allow Composition authors to specify what resources they want, e.g. by name, by
+label selector, based on fields from the Composite Resource, or even
+another Composed resource someday... The API would need to be at least as
+expressive, and therefore complex, as what we have today for
+`EnvironmentConfigs` at `spec.environment.environmentConfigs`, or even as that
+of [Kyverno][kyverno-context].
+
+**Option 1** is clearly more flexible, as it would allow Function authors to
+expose whatever API to Composition authors through inputs, or just decide it
+according to whatever logic they want. However, this will obviously require at
+least one more round-trip to get the requested resources from Crossplane to the
+Function.
```

Comment:
Nothing comes for free. Flexibility is the obvious benefit, but another drawback is distributing the logic to multiple places. Instead of being able to say, "this pipeline in my composition requires values from the EnvironmentConfig named foo" in a single place (e.g. composition.yaml), now I need to do that in two distinct places, "my composition can request EnvironmentConfigs" and in my function "I need EnvironmentConfig named foo". Former would work in most cases but not all unless we commit to a complex API and the latter complicates the overall system. However, reusable functions could alleviate this and I agree that starting with the flexible option makes more sense and better aligns with the direction that Crossplane is heading with the introduction of functions.